### PR TITLE
ENH Add generic types

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Description

Adds generic types (mostly return types - we don't need generic param typing for now since we're not really using static analysis).

There are also a few general corrections to PHPDocs that I noticed along the way (e.g. adding `|null` when the method is returning a null value.

There are some cases where either the return type or the whole PHPDoc was duplicated from the parent class - in those cases I've simply removed the duplication.

### Caveats
1. `DataObject::get()` and `DataObject::get_one()` give weird results in some specific scenarios. 
    - If using `DataObject:get(MyClass::class)`, it won't know about the `MyClass` class and just give you a `DataList<DataObject>` - which is correct, but imprecise. Same with `get_one()` with the same syntax.
    - If using `MyClass::get(AnotherClass::class)`, it won't know about the `AnotherClass` and will give you a `DataList<MyClass>` - which is NOT correct. I think this is okay since doing that is a bad code smell anyway - people should just not do that. Probably in a future release we should explicitly disallow that syntax. Same with `get_one()` with the same syntax.
    - We could use a conditional return type here which would be useful for static analysis, but that's not supported in VSCode using intellephense so we shouldn't do that.
1. Because `ArrayList` will wrap an associative array with `ArrayData` in its iterator (but not in `First()`, `Last()`, etc), the typing for the iterator is wrong if the arraylist was created with associative arrays. Ideally people should be wrapping their associative arrays in `ArrayData` anyway because of the existing inconsistency with what `ArrayList` returns.
1. Because `ArrayList` can accept values of mixed types (you could have a combination of `ArrayData`, `DataObject`, and some custom class for example), the return values might not always be reported correctly, especially if adding objects to the list _after_ instantiation. This is acceptable since that sort of usage is less common - and they'd already have to use `@var` or similar in their IDE before this PR.
1. When passing a service name that isn't a direct FQCN to the injector (e.g. `Injector::inst()->get(LoggerInterface::class . '.error');`), there is no way for an IDE to determine the correct class name. In the future we may want to use conditional return types to try to resolve this sort of thing, but as of writing those aren't as widely supported in IDEs as non-conditional generic return types are.
1. For the extension generics to really be useful, subclasses will need to use the appropriate `@extends` annotation. I've raised [a separate card](https://github.com/silverstripe/.github/issues/181) to add a CI rule for this so we can at least be consistent with it in supported modules.
1. `SapphireTest::objFromFixture()` can technically take a table name as the first argument.... but that's probably never done in practice. I'm 99% sure in that case most IDEs would correctly still recognise that it's returning `DataObject` anyway so should be fine.

## Manual testing steps

Check that things work as expected in your IDE.

## Issues

- https://github.com/silverstripe/silverstripe-framework/issues/11066

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] Tests aren't necessary for this change (we may add static analysis later, separately)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

File: src/Core/Injector/Injector.php [hunk 1]
```diff
@@ -966,12 +966,13 @@ public function unregisterObjects($types)
...
      *
-     * @param string $name The name of the service to retrieve. If not a registered
+     * @template T of object
+     * @param class-string<T> $name The name of the service to retrieve. If not a registered
      * service, then a class of the given name is instantiated
      * @param bool $asSingleton If set to false a new instance will be returned.
      * If true a singleton will be returned unless the spec is type=prototype'
      * @param array $constructorArgs Args to pass in to the constructor. Note: Ignored for singletons
-     * @return mixed Instance of the specified object
+     * @return T|mixed Instance of the specified object
```
Comment (emteknetnz) [inline src/Core/Injector/Injector.php:975]: `T|mixed`?

Shouldn't this be either `T` or `mixed` but not both?

Applies to all instances of `@return T|mixed`

File: src/ORM/ArrayList.php [hunk 2]
```diff
@@ -596,8 +588,7 @@ public function canFilterBy($by)
      * Find the first item of this list where the given key = value
      *
      * @param string $key
-     * @param string $value
-     * @return mixed
+     * @param mixed $value
```
Comment (emteknetnz) [inline src/ORM/ArrayList.php:591]: `@return T|null`?

File: src/Core/Extension.php [hunk 3]
```diff
@@ -13,6 +13,8 @@
  * Every object instance gets its own set of extension instances,
  * meaning you can set parameters specific to the "owner instance"
  * in new Extension instances.
+ *
+ * @template T of object
```
Comment (emteknetnz) [inline src/Core/Extension.php:17]: Through this PR there's a mixture of
`@template T of object`
`@template T`

Are they functionally the same? If so should we just standardise on one (probably `@template T`)?

File: src/ORM/DataList.php [hunk 4]
```diff
@@ -677,8 +681,6 @@ protected function isValidRelationName($field)
      *
      * @param string|array
      * @param string [optional]
-     *
-     * @return $this
```
Comment (emteknetnz) [inline src/ORM/DataList.php:681]: `@return static<T>`

File: src/ORM/DataList.php [hunk 5]
```diff
@@ -822,8 +824,6 @@ public function rightJoin($table, $onClause, $alias = null, $order = 20, $parame
     /**
      * Return an array of the actual items that this DataList contains at this stage.
      * This is when the query is actually executed.
-     *
-     * @return array
```
Comment (emteknetnz) [inline src/ORM/DataList.php:826]: `@return array<T>`?

Same with `toNestedArray()` below?

File: src/ORM/DataList.php [hunk 6]
```diff
@@ -1548,11 +1545,9 @@ public function sum($fieldName)
 
 
     /**
-     * Returns the first item in this DataList (instanceof DataObject)
+     * Returns the first item in this DataList
      *
      * The object returned is not cached, unlike {@link DataObject::get_one()}
-     *
-     * @return DataObject|null
```
Comment (emteknetnz) [inline src/ORM/DataList.php:1555]: `@return T|null`?

Same with `last()` and `find()`?

File: src/ORM/DataList.php [hunk 7]
```diff
@@ -1625,12 +1617,6 @@ public function setQueriedColumns($queriedColumns)
         });
     }
 
-    /**
-     * Filter this list to only contain the given Primary IDs
-     *
-     * @param array $ids Array of integers
-     * @return $this
```
Comment (emteknetnz) [inline src/ORM/DataList.php:1632]: Why is this docblock being removed?  And shouldn't the return type be `static<T>`?

File: src/ORM/DataList.php [hunk 8]
```diff
@@ -1640,9 +1626,6 @@ public function byIDs($ids)
      * Return the first DataObject with the given ID
      *
      * The object returned is not cached, unlike {@link DataObject::get_by_id()}
-     *
-     * @param int $id
-     * @return DataObject|null
```
Comment (emteknetnz) [inline src/ORM/DataList.php:1645]: Why is this being removed? Return type should be `T|null`?

File: src/ORM/DataObject.php [hunk 9]
```diff
@@ -742,9 +739,10 @@ public function setClassName($className)
      * there is no record, or the record has no ID. In this case, we do have an ID but
      * we still need to repopulate the defaults.
      *
-     * @param string $newClassName The name of the new class
+     * @template T of DataObject
```
Comment (emteknetnz) [inline src/ORM/DataObject.php:742]: Does defining `@template` on `newClassInstance()` mean that only `T` is only used if the DataObject was instantiated using this method? i.e. there is no `@template` if I go `$obj = new MyDataObject();`?

Is this why methods like `DataObject::update()` return `static` instead of `T`? Or is that simply because `static` will always be correct so no need to be cute and `@return T`?

**Review by emteknetnz — CHANGES_REQUESTED** [review]
There's a lot to peer review here so all of the peer review will take multiple peer review sessions to get through.  I got about halfway through and got up to (but haven't reviewed) EagerLoadedList

File: src/Core/Injector/Injector.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Core/Injector/Injector.php:975]: `T|mixed` does mean "`T` or `mixed` but not both" (that's what union types explicitly mean).... so I'm not sure what you mean?

File: src/ORM/ArrayList.php (hunk 2 above)
Comment (GuySartorelli) [reply src/ORM/ArrayList.php:591]: I'm not sure what you're asking, please be more verbose. If you're asking "Why does this not include `@return T|null`?" then the answer is that this method and its signature and phpdoc typehints are inherited from the `Filterable` interface - there's no need to repeat the return typehint here since it doesn't differ from what's already declared in the PHPDoc on the interface's method declaration.

This is the same for any other methods which inherit methods from an interface or superclass where the typehint doesn't need to be repeated.

File: src/Core/Extension.php (hunk 3 above)
Comment (GuySartorelli) [reply src/Core/Extension.php:17]: `@template t of object` says "`T` will always be an object, of some kind".
`@template t` (with no `of` part) means "This will be something, but we don't know what yet"

We should be specific when we can.

File: src/ORM/DataList.php (hunk 4 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php:681]: https://github.com/silverstripe/silverstripe-framework/pull/11108#discussion_r1451866912

File: src/ORM/DataList.php (hunk 5 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php:826]: https://github.com/silverstripe/silverstripe-framework/pull/11108#discussion_r1451866912

File: src/ORM/DataList.php (hunk 6 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php:1555]: https://github.com/silverstripe/silverstripe-framework/pull/11108#discussion_r1451866912

File: src/ORM/DataList.php (hunk 7 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php:1632]: This docblock doesn't differ at all from the superclass or interface it's inheriting the method from. No need to repeat it.
https://github.com/silverstripe/silverstripe-framework/pull/11108#discussion_r1451866912

File: src/ORM/DataList.php (hunk 8 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php:1645]: https://github.com/silverstripe/silverstripe-framework/pull/11108#discussion_r1451866912

File: src/ORM/DataObject.php (hunk 9 above)
Comment (GuySartorelli) [reply src/ORM/DataObject.php:742]: > Does defining `@template` on `newClassInstance()` mean that only `T` is only used if the DataObject was instantiated using this method? i.e. there is no `@template` if I go `$obj = new MyDataObject();`?

I don't understand your question, but it seems to revolve around how the `@template T` used in this method's PHPDoc differs from the one in the class PHPDoc, so I'll try to explain that.

The class PHPDoc's `@template T` applies to the whole class, unless otherwise stated. Everywhere that has `T` as a type refers to the `T` in the class phpdoc unless there's a `@template T` in a narrower scope.

Using `@template T` here in the scope of this method's PHPDoc gives us a _new, separate_ `T` type. This `T` type will resolve to whatever is passed in to the `$newClassName` argument, since that has a typehint of `class-string<T>`

In other words, even if this is a `DataList<Member>`, when calling this method, the return type will _always_ resolve to whatever class is passed into this method. So you end up with the equivalent of this:

```php
/** @var DataList<Member> $list */
$list = Member::get();
/** @var Group $record */
$record = $list->newClassInstance(Group::class);
```

If that hasn't answered your question, please try rewording it.

Comment (GuySartorelli) [reply src/ORM/DataObject.php:742]: > Is this why methods like `DataObject::update()` return `static` instead of `T`? Or is that simply because `static` will always be correct so no need to be cute and `@return T`?

`DataObject::update()` doesn't have any way to identify what `T` would be - so it doesn't make sense to try to use generics there. If the `update()` method took a classname variable and always returned the type of object that classname represents, we could use generics there - but since there's nothing like that, we just use `static`.

File: src/Core/Injector/Injector.php (hunk 1 above)
Comment (Cambis) [reply src/Core/Injector/Injector.php:975]: I reckon use:

```php
@param class-string<T>|string $name
...
@return T|mixed
```

Which would imply that if a class-string is used you get a type of `T`, but if you use a regular string you'd get a `mixed` type.

Comment (emteknetnz) [reply src/Core/Injector/Injector.php:975]: Either have `@return T` or `@return mixed` but not `@return T|mixed`

`mixed` means "everything" (even null), so T is a subset of `mixed`.  This means there's no point having `T|mixed` since it's essentially just `mixed`.

However I'm not sure why we have `mixed` here since this should just be returning `T` the whole time right unless we throw an exception?

File: src/ORM/ArrayList.php (hunk 2 above)
Comment (emteknetnz) [reply src/ORM/ArrayList.php:591]: OK, I think you meant `SS_List` interface

This PR is a bit of a nightmare since it essentially has lots of unrelated refactoring in it (changing docblocks not related to generics)

File: src/ORM/ArrayList.php [hunk 10]
```diff
@@ -45,13 +51,12 @@ class ArrayList extends ViewableData implements SS_List, Filterable, Sortable, L
...
      *
-     * @var array
+     * @var array<array-key, T>
      */
     protected $items = [];
 
     /**
-     *
-     * @param array $items - an initial array to fill this object with
+     * @param array<array-key, T> $items - an initial array to fill this object with
```
Comment (emteknetnz) [inline src/ORM/ArrayList.php:59]: From the description:

- Because ArrayList can accept values of mixed types (you could have a combination of ArrayData, DataObject, and some custom class for example), the return values might not always be reported correctly, especially if adding objects to the list after instantiation. This is acceptable since that sort of usage is less common - and they'd already have to use @var or similar in their IDE before this PR.

What actually sets the `@template T` when you pass in a bunch of mixed type arguments in the constructor? The first value? The last value? (it changes on each item added)? Or do you need to call `setDataClass()` directly?

> This is acceptable since that sort of usage is less common - and they'd already have to use @var or similar in their IDE before this PR.

> Ideally people should be wrapping their associative arrays in ArrayData anyway because of the existing inconsistency with what ArrayList returns.

That's debatable that it's acceptable? I think we're better off just not trying to force using generics in ArrayList if we really should just be using mixed here instead, rather that saying it's the users fault for doing it wrong (despite it being perfectly allowable)

File: src/ORM/ArrayList.php (hunk 2 above)
Comment (emteknetnz) [reply src/ORM/ArrayList.php:591]: OK, I think you meant `SS_List` interface

Kind feels like like there's loads of refactoring in this PR unrelated to generics specifically.  Since it's a bit of nightmare validating all of this and because docblocks types are known to be innaccurate I'm going to largely just ignore alot of those types of docblock changes where things were removed

That said - doesn't that now mean the interface now has the wrong docblock type? i.e. `SS_List::find()` should be `@return T|null`

We don't need to be too concerned about semver since docblocks aren't real so we're not going to break anyones site by changing them

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Core/Injector/Injector.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Core/Injector/Injector.php:975]: > Either have `@return T` or `@return mixed` but not `@return T|mixed`
>
> `mixed` means "everything" (even null), so T is a subset of `mixed`.  This means there's no point having `T|mixed` since it's essentially just `mixed`.

While `mixed` does mean "everything", the whole point of this is to tell the IDE what it should be expecting, so you can (for example) start typing method names and the IDE will tell you what methods are available.

e.g:
```php
Injector::inst()->get(DataObject::class)->val
```
Typing the above, my IDE will tell me there's a `validate()` method - but if we just hinted the return type was `mixed`, it can't do that.

> However I'm not sure why we have `mixed` here since this should just be returning `T` the whole time right unless we throw an exception?

The `mixed` here is because sometimes service names aren't FQCN - e.g. for the error logging handler:
```php
Injector::inst()->get(LoggerInterface::class . '.errorhandler');
```
There's no such class as `Psr\Log\LoggerInterface.errorhandler` - so the `|mixed` handles this scenario.

Cam's suggestion is a good one, as it provides better information to the developer as to what scenario results in the `mixed` return type. I'll make that change.

Comment (GuySartorelli) [reply src/Core/Injector/Injector.php:975]: Note that quite often you'll pass FQCN in so the generic typehint is useful to add - and when you _don't_ pass a FQCN you'll need to add a `@var` comment, the same as before this change was made. So this is a net positive change.

File: src/ORM/ArrayList.php (hunk 2 above)
Comment (GuySartorelli) [reply src/ORM/ArrayList.php:591]: > That said - doesn't that now mean the interface now has the wrong docblock type? i.e. `SS_List::find()` should be `@return T|null`

It is `@return T|null`: https://github.com/silverstripe/silverstripe-framework/pull/11108/files#diff-55afe76c21c7906f822298c339509216643b0b56d33607ecf43e93bff6617b34R78

File: src/ORM/ArrayList.php (hunk 10 above)
Comment (GuySartorelli) [reply src/ORM/ArrayList.php:59]: > What actually sets the `@template T` when you pass in a bunch of mixed type arguments in the constructor? The first value? The last value? (it changes on each item added)? Or do you need to call `setDataClass()` directly?

Don't forget that this is static analysis - it's not actually executing the code. So if your IDE knows what's in the array, it'll tell you `T` is a union of those types (e.g. you might have `ArrayList<stdClass|ArrayData|Member>`. If your IDE doesn't know what's in the array, then `T` will be non-determinate, which resolves back to `mixed`.

And besides all that, most of the time when instantiating a new arraylist you'll use `ArrayList::create()`, so your IDE won't know what's being passed to the constructor anyway, so you'll see `T` as `mixed`.

In any case, even if that didn't happen I'm not sure we need to worry too much about it. As I mentioned in the description, to get the correct type when your list contains mixed types you already have to use `@var` to tell your IDE what to expect, and that doesn't change with this PR.

This PR gives you correct types when you use only a specific type - and from what I can tell it _doesn't_ deteriorate the DX when using mixed types.

## The main benefit of generics in `ArrayList`

I suspect this will be by far the more common scenario - having generics in `Arraylist` allows you to use `@return ArrayList<SomeType>`, so that it's clear what an arraylist contains when it's returned from a method you've called.

With that in mind, do you think we should revert this change or leave it in? I can go either way.

File: src/Core/Injector/Injector.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Core/Injector/Injector.php:975]: I've made the change Cam suggested.

File: src/ORM/ArrayList.php (hunk 10 above)
Comment (emteknetnz) [reply src/ORM/ArrayList.php:59]: > And besides all that, most of the time when instantiating a new arraylist you'll use ArrayList::create(), so your IDE won't know what's being passed to the constructor anyway, so you'll see T as mixed.

How come this doesn't work?  I tested this locally and these two behave differently:

```php
// Type hints MyDataObject
$list = new ArrayList([
    new MyDataObject()
]);
// Won't type hint MyDataObject
$list = ArrayList::create([
    new MyDataObject()
]);
foreach($list as $obj) {
    $obj->mymethod; // defined on MyDataObject, only new ArrayList() will type hint it
}
```

> I suspect this will be by far the more common scenario - having generics in Arraylist allows you to use `@return` ArrayList<SomeType>

By this do you mean project code will add this docblock return type to a method defined in the project?

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/ORM/ArrayList.php (hunk 10 above)
Comment (GuySartorelli) [reply src/ORM/ArrayList.php:59]: > > And besides all that, most of the time when instantiating a new arraylist you'll use ArrayList::create(), so your IDE won't know what's being passed to the constructor anyway, so you'll see T as mixed.
>
> How come this doesn't work?

`create()` isn't the constructor, so it doesn't have the generic typehints available. Static analysis tools could probably be made to understand what's going on there but IDEs aren't that smart.

> > I suspect this will be by far the more common scenario - having generics in Arraylist allows you to use `@return` ArrayList
>
> By this do you mean project code will add this docblock return type to a method defined in the project?

Projects _can_ add those, but I honestly don't expect them to. I mostly mean that _we_ will use the generic return types (e.g. `@return ArrayList<Member>`), and projects can then benefit from those by their IDEs correctly identifying the contents of the returned lists.

File: src/ORM/ListDecorator.php [hunk 11]
```diff
@@ -10,22 +10,34 @@
...
+ * @implements Sortable<T>
+ * @implements Filterable<T>
+ * @implements Limitable<T>
  */
 abstract class ListDecorator extends ViewableData implements SS_List, Sortable, Filterable, Limitable
 {
-
     /**
-     * @var SS_List
+     * @var TList<T>
```
Comment (emteknetnz) [inline src/ORM/ListDecorator.php:24]:
```suggestion
     * @var TList<T>|SS_List&Sortable&Filterable&Limitable
```
This is to resolve vscode complaining about strongly type `getList()` returning a `TList<T>` instead of a `SS_List&Sortable&Filterable&Limitable`

File: src/ORM/PaginatedList.php [hunk 12]
```diff
@@ -13,10 +13,13 @@
 
 /**
  * A decorator that wraps around a data list in order to provide pagination.
+ *
+ * @template TList
```
Comment (emteknetnz) [inline src/ORM/PaginatedList.php]:
```suggestion
 * @template TList of SS_List
```
?

File: src/Security/InheritedPermissionsExtension.php [hunk 13]
```diff
@@ -14,6 +14,8 @@
  * @method ManyManyList<Member> EditorMembers()
  * @method ManyManyList<Group> ViewerGroups()
  * @method ManyManyList<Member> ViewerMembers()
+ *
+ * @extends DataExtension<DataObject&static>
```
Comment (emteknetnz) [inline src/Security/InheritedPermissionsExtension.php]:
```suggestion
 * @extends DataExtension<File|SiteTree>
```

Based on `@extends DataExtension<Member|Group>` in InheritedPermissionFlusher.php a couple of files up, I assume that the syntax is `@extends [class that is extended]<[Classes the extension is known to be applied to]>`.  Not sure how `&static` would be applicable here

File: src/View/ViewableData.php [hunk 14]
```diff
@@ -644,10 +645,13 @@ public function getXMLValues($fields)
      * This is useful so you can use a single record inside a <% control %> block in a template - and then use
      * to access individual fields on this object.
      *
+     * @deprecated 5.2.0 Will be removed without equivalent functionality
+     *
      * @return ArrayIterator
      */
     public function getIterator(): Traversable
     {
+        Deprecation::notice('5.2.0', 'Will be removed without equivalent functionality');
```
Comment (emteknetnz) [inline src/View/ViewableData.php:654]: Shouldn't be deprecating things in the PR, it's very unrelated.  Why are we deprecating this here?

File: src/Forms/GridField/GridFieldConfig.php [hunk 15]
```diff
@@ -111,7 +111,7 @@ public function removeComponentsByType($types)
     }
 
     /**
-     * @return ArrayList Of GridFieldComponent
+     * @return ArrayList<GridFieldComponent> Of GridFieldComponent
```
Comment (emteknetnz) [inline src/Forms/GridField/GridFieldConfig.php]:
```suggestion
     * @return ArrayList<GridFieldComponent>
```

File: src/Forms/GridField/GridFieldConfig.php [hunk 16]
```diff
@@ -124,8 +124,9 @@ public function getComponents()
     /**
      * Returns all components extending a certain class, or implementing a certain interface.
      *
-     * @param string $type Class name or interface
-     * @return ArrayList Of GridFieldComponent
+     * @template T of GridFieldComponent
+     * @param class-string<T> $type Class name or interface
+     * @return ArrayList<T> Of GridFieldComponent
```
Comment (emteknetnz) [inline src/Forms/GridField/GridFieldConfig.php]:
```suggestion
     * @return ArrayList<T>
```

File: src/ORM/Filterable.php [hunk 17]
```diff
@@ -11,6 +11,9 @@
  * @see SS_List
  * @see Sortable
  * @see Limitable
+ *
+ * @template T
+ * @implements SS_List<T>
```
Comment (emteknetnz) [inline src/ORM/Filterable.php]: Should the be `@extends` instead of `@implements`?

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Security/InheritedPermissionsExtension.php (hunk 13 above)
Comment (GuySartorelli) [reply src/Security/InheritedPermissionsExtension.php]: This could be applied in project code to any arbitrary model, so we shouldn't specify `File|SiteTree` here.

> Not sure how `&static` would be applicable here

`&static` says "And also it has the methods/properties available in this class". I've done that for any extension subclasses where we're referencing extension methods on the owner. Looks like I might have accidentally added it to this one as well, so I'll remove it from here.

But for example in `Hierarchy` we have `$this->owner->stageChildren()`, where `stageChildren()` is a public method defined in the `Hierarchy` extension itself.

File: src/View/ViewableData.php (hunk 14 above)
Comment (GuySartorelli) [reply src/View/ViewableData.php:654]: It's not unrelated. I thought I had added a comment to the PR about this but it looks like I forgot to.

This is being deprecated because:
- It's not actually needed anymore - the `<% control %>` template block was removed in CMS 3, and that was the only reason this method is here at all.
- having the `getIterator()` here causes problems with identifying the correct type for `T` being iterated through from `getIterator()` of subclasses in some IDEs. This was brought up in the original POC [in this issue](https://github.com/silverstripeltd/silverstripe-framework-generics/issues/19). One solution we considered was adding generic typing to `ViewableData` but that really is nonsensical outside of the context of lists, and _most_ `ViewableData` isn't a list.

File: src/Forms/GridField/GridFieldConfig.php (hunk 15 above)
Comment (GuySartorelli) [reply src/Forms/GridField/GridFieldConfig.php]: Oops yup. Good catch. Fixed.

File: src/Forms/GridField/GridFieldConfig.php (hunk 16 above)
Comment (GuySartorelli) [reply src/Forms/GridField/GridFieldConfig.php]: Fixed.

File: src/ORM/Filterable.php (hunk 17 above)
Comment (GuySartorelli) [reply src/ORM/Filterable.php]: Oops yup. Fixed.

File: src/Security/InheritedPermissionsExtension.php (hunk 13 above)
Comment (GuySartorelli) [reply src/Security/InheritedPermissionsExtension.php]: Removed `&static` from this one

File: src/ORM/PaginatedList.php (hunk 12 above)
Comment (GuySartorelli) [reply src/ORM/PaginatedList.php]: Ahh yup good catch. Fixed.

File: src/ORM/ListDecorator.php (hunk 11 above)
Comment (GuySartorelli) [reply src/ORM/ListDecorator.php:24]: `T` _is_ `SS_List&Sortable&Filterable&Limitable`.

I assume the complaint you're talking about is this red line here?
image

If so, I'm inclined to ignore that. We know it's correct. When you call `getList()` on a list and look at what it's resolving to, it's clear vscode knows it's correct, too. Try this out and see what the type of `$y` is.
```php
/** @var ListDecorator $x */
$y = $x->getList();
```

This looks more like a bug with vscode or with intellephense to me, and I don't think we should cater our PHPDoc to bugs in IDE plugins.

Comment (GuySartorelli) [reply src/ORM/ListDecorator.php:24]: That said, I will make the change if you insist

Comment (emteknetnz) [reply src/ORM/ListDecorator.php:24]: Yeah you're probably right, just omit my recommendation

File: src/View/ViewableData.php (hunk 14 above)
Comment (emteknetnz) [reply src/View/ViewableData.php:654]: OK.  Add it to the 5.2.0 changelog that it's being deprecated I suppose then.

**Review by emteknetnz — APPROVED** [review]
